torchaudio_contrib/layers.py

```diff
@@ -1,142 +1,241 @@
-
-    Its implemented as a layer so that the computation can be faster (done dynamically
-    on GPU) and no need to store the transforms. More information:
-        - https://github.com/keunwoochoi/kapre
-        - https://arxiv.org/pdf/1706.05781.pdf
+    Extension of nn.Module that removes buffers
+    from state_dict.
 
     Args:
-     * hop: int > 0
-       -  Hop length between frames in sample,  should be <= n_fft.
-       -  Default: None (in which case n_fft // 4 is used)
-     * n_fft: int > 0 
-       -  Size of the fft.
-       -  Default: 2048
-     * pad: int >= 0
-       -  Amount of two sided zero padding to apply.
-       -  Default: 0
-     * window: torch.Tensor,
-       -  Windowing used in the stft.
-       -  Default: None (in which case torch.hann_window(n_fft) is used)
-     * sr: int > 0
-       -  Sampling rate of the audio signal. This may not be the same in all samples (?)
-       -  Default: 44100
-     * power: int = 1,2,None
-       -  Power to normalize the spectrogram to, make None to work with complex stft
-       -  Default: 1
-     * kwargs: 
-       -  Any named arguments to be passed to the stft
+        buffers (dict): mapping of names and tensors to be registered.
+    """
 
+    def __init__(self, buffers):
+        super(_ModuleNoStateBuffers, self).__init__()
+        for k, v in buffers.items():
+            self.register_buffer(k, v)
+
+    def state_dict(self, destination=None, prefix='', keep_vars=False):
+        ret = super(_ModuleNoStateBuffers, self).state_dict(
+            destination, prefix, keep_vars)
+        for k in self._buffers.keys():
+            del ret[prefix + k]
+        return ret
+
+
+class STFT(_ModuleNoStateBuffers):
     """
+    Compute the stft transform of a multi-channel signal or
+    batch of multi-channel signals.
 
-    def __init__(self, hop=None, n_fft=2048, pad=0, window=None, sr=44100, **kwargs):
+    Args:
 
-        super(Spectrogram, self).__init__()
+        n_fft (int): FFT window size. Defaults to 2048.
+        hop_length (int): Number audio of frames between stft columns.
+            Defaults to n_fft // 4.
+        len_win (int): Size of stft window. Defaults to n_fft.
+        window (Tensor): 1-D tensor. Defaults to Hanning Window
+            of size len_win.
+        pad (int): Amount of padding to apply to signal. Defaults to 0.
+        pad_mode: padding method (see torch.nn.functional.pad).
+            Defaults to "reflect".
+        **kwargs: Other torch.stft parameters, see torch.stft for more details.
 
-        if window is None:
-            window = torch.hann_window(n_fft)
+    """
 
-        self.window = self._build_window(window)
-        self.hop = n_fft // 4 if hop is None else hop
-        self.n_fft = n_fft
-        self.pad = pad
+    def __init__(self, n_fft=2048, hop_length=None, len_win=None,
+                 window=None, pad=0, pad_mode="reflect", **kwargs):
```

Comment:
Yeah exactly, and I'm not too sure. I'd be more inclined to follow `torch.stft` and let it handle the padding but perhaps it was included in torchaudio for a use case.

Reply:
> perhaps it was included in torchaudio for a use case.

I'll ask in the slack channel.

Reply:
> I'll ask in the slack channel.

@dhpollack replied: "There is no specific reason. It was just the easiest way to do padding."

There's one thing we might want to change in comparison to `torch.stft`: When the window size is even (which is quite common), `torch.stft` with `center=True` still does symmetric padding, leading to an output that can have one frame too much (i.e., has its first frame centered on the first sample, but the last frame centered on the sample after the last sample, where "centered" is of course ambiguous in an even window).

Reply:
Ok center being ambiguous for an even sized window makes sense, but I'm not too sure I follow the extra frame problem.

Given the padding when `center=True` the output dimension of the stft columns will be:
```
(input_length + 2*pad - n_fft + hop_length) // hop_length
```
With `pad=n_fft//2`. So the only way I see and extra frame being added is if `input_length%hop_length==0`? 

So in cases like:
```
In [124]: (256*123 + 2*(2047//2) - 2047 + 256)//256
Out[124]: 123

In [125]: (256*123 + 2*(2048//2) - 2048 + 256)//256
Out[125]: 124

In [126]: (256*123 + 2*(2049//2) - 2049 + 256)//256
Out[126]: 123
```
I may have misunderstood though.

Reply:
> I'm not too sure I follow the extra frame problem.

Let's assume we have 4 samples, a window size of 4, and a stride of 2, then with PyTorch's padding we get:
```
00ssss00
wwww
  wwww
    wwww
```
If we define the first window to be centered on the first sample, then the third window is already centered on the zero-padding (because it's positioned 4 samples later). To avoid this, we should pad with 2 zeros on the left and 1 zero on the right.

Reply:
..or, both the first window and third window are centered at the exact boundaries? I'm not sure why it's a problem. 

> To avoid this, we should pad with 2 zeros on the left and 1 zero on the right.

and how is the third window applied..?

Reply:
Btw, 
> @dhpollack replied: "There is no specific reason. It was just the easiest way to do padding."

So there's no reason for non-zero padding that we know, hence I think we can totally move the padding into `torch.stft` by now.

Reply:
@keunwoochoi I was talking about why I used the padding function in that comment. The reason I used non-zero passing was that I had problems if one of the windows was all zeros. But I believe this was before the stft function existed.

Reply:
@dhpollack Thanks for confirming it. 👍👍